src/ChunkedStreamDecoder.php

```diff
@@ -0,0 +1,105 @@
+<?php
+
+namespace React\HttpClient;
+
+use Evenement\EventEmitterTrait;
+use React\Stream\DuplexStreamInterface;
+use React\Stream\Util;
+
+class ChunkedStreamDecoder
+{
+    const CRLF = "\r\n";
+
+    use EventEmitterTrait;
+
+    /**
+     * @var string
+     */
+    protected $buffer = '';
+
+    /**
+     * @var int
+     */
+    protected $remainingLength = 0;
+
+    /**
+     * @var bool
+     */
+    protected $nextChunkIsLength = true;
+
+    /**
+     * @var DuplexStreamInterface
+     */
+    protected $stream;
+
+    /**
+     * @param DuplexStreamInterface $stream
+     */
+    public function __construct(DuplexStreamInterface $stream)
+    {
+        $this->stream = $stream;
+        $this->stream->on('data', array($this, 'handleData'));
+        Util::forwardEvents($this->stream, $this, [
+            'error',
+            'end',
+        ]);
+    }
+
+    public function handleData($data)
+    {
+        $this->buffer .= $data;
+
+        do {
+            $this->iterateBuffer();
+        } while (strlen($this->buffer) > 0 && strpos($this->buffer, static::CRLF) !== false);
+    }
+
+    protected function iterateBuffer()
+    {
+        if ($this->nextChunkIsLength) {
+            $this->nextChunkIsLength = false;
+            $this->remainingLength = hexdec(substr($this->buffer, 0, strpos($this->buffer, static::CRLF)));
```

Comment:
Should test for [chunk extensions](https://en.wikipedia.org/wiki/Chunked_transfer_encoding#Format) here.

>If chunk extensions are provided, the chunk size is terminated by a semicolon and followed by the parameters, each also delimited by semicolons. Each parameter is encoded as an extension name followed by an optional equal sign and value. These parameters could be used for a running message digest or digital signature, or to indicate an estimated transfer progress, for instance.